src/ray/gcs/gcs_server/gcs_placement_group_manager.h

```diff
@@ -78,5 +78,93 @@ class GcsPlacementGroup {
+class GcsPlacementGroupManager : public rpc::PlacementGroupInfoHandler {
+ public:
+  /// Create a GcsPlacementGroupManager
+  ///
+  /// \param io_context The event loop to run the monitor on.
+  /// \param scheduler Used to schedule placement group creation tasks.
+  /// \param gcs_table_storage Used to flush placement group data to storage.
+  explicit GcsPlacementGroupManager(
+      boost::asio::io_context &io_context,
+      std::shared_ptr<GcsPlacementGroupSchedulerInterface> scheduler,
+      std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage);
+
+  ~GcsPlacementGroupManager() = default;
+
+  void HandleCreatePlacementGroup(const rpc::CreatePlacementGroupRequest &request,
+                                  rpc::CreatePlacementGroupReply *reply,
+                                  rpc::SendReplyCallback send_reply_callback) override;
+
+  /// Register placement_group asynchronously.
+  ///
+  /// \param request Contains the meta info to create the placement_group.
+  /// \param callback Will be invoked after the placement_group is created successfully or
+  /// be invoked immediately if the placement_group is already registered to
+  /// `registered_placement_groups_` and its state is `ALIVE`. The callback will not be
+  /// called in this case.
+  Status RegisterPlacementGroup(const rpc::CreatePlacementGroupRequest &request,
+                                RegisterPlacementGroupCallback callback);
+
+  /// Schedule placement_groups in the `pending_placement_groups_` queue.
+  void SchedulePendingPlacementGroups();
+
+  /// Get the placement_group ID for the named placement_group. Returns nil if the
+  /// placement_group was not found.
+  /// \param name The name of the  placement_group to look up.
+  /// \returns PlacementGroupID The ID of the placement_group. Nil if the
+  /// placement_group was not found.
+  PlacementGroupID GetPlacementGroupIDByName(const std::string &name);
+
+  /// Handle placement_group creation task failure. This should be called when scheduling
+  /// an placement_group creation task is infeasible.
+  ///
+  /// \param placement_group The placement_group whose creation task is infeasible.
+  void OnPlacementGroupCreationFailed(std::shared_ptr<GcsPlacementGroup> placement_group);
+
+  /// Handle placement_group creation task success. This should be called when the
+  /// placement_group creation task has been scheduled successfully.
+  ///
+  /// \param placement_group The placement_group that has been created.
+  void OnPlacementGroupCreationSuccess(
+      std::shared_ptr<GcsPlacementGroup> placement_group);
+
+ private:
+  /// Schedule another tick after a short time.
+  void ScheduleTick();
+
+  /// Callbacks of placement_group registration requests that are not yet flushed.
+  /// This map is used to filter duplicated messages from a Driver/Worker caused by some
+  /// network problems.
+  absl::flat_hash_map<PlacementGroupID, std::vector<RegisterPlacementGroupCallback>>
```

Comment:
If it is for the filtering, why do we make it a vector? Isn't a single entry sufficient?

Reply:
> If it is for the filtering, why do we make it a vector? Isn't a single entry sufficient?

That's because we do not know which callback is the latest as the GRPC an not make a promise of the order. So we need cache all of the duplicated items and invoke them once the placement group is created.

Reply:
So, that says the RegisterPlacementGroupCallback must be idempotent and being independent from ordering right? Can you write some comments to explain that?